packages/aws-cdk/lib/api/garbage-collector.ts

```diff
@@ -0,0 +1,484 @@
+  private garbageCollectEcrAssets: boolean;
+  private permissionToDelete: boolean;
+  private permissionToTag: boolean;
+  private bootstrapStackName: string;
+  private maxWaitTime: number;
+
+  public constructor(readonly props: GarbageCollectorProps) {
+    this.garbageCollectS3Assets = ['s3', 'all'].includes(props.type);
+    this.garbageCollectEcrAssets = ['ecr', 'all'].includes(props.type);
+
+    this.permissionToDelete = ['delete-tagged', 'full'].includes(props.action);
+    this.permissionToTag = ['tag', 'full'].includes(props.action);
+    this.maxWaitTime = props.maxWaitTime ?? 60000;
+
+    print(chalk.white(this.permissionToDelete, this.permissionToTag, props.action));
+
+    this.bootstrapStackName = props.bootstrapStackName ?? DEFAULT_TOOLKIT_STACK_NAME;
+
+    // TODO: ECR garbage collection
+    if (this.garbageCollectEcrAssets) {
+      throw new Error('ECR garbage collection is not yet supported');
+    }
+  }
+
+  /**
+   * Perform garbage collection on the resolved environment.
+   */
+  public async garbageCollect() {
+    print(chalk.black(this.garbageCollectS3Assets));
+    // SDKs
+    const sdk = (await this.props.sdkProvider.forEnvironment(this.props.resolvedEnvironment, Mode.ForWriting)).sdk;
+    const cfn = sdk.cloudFormation();
+    const s3 = sdk.s3();
+
+    const qualifier = await this.bootstrapQualifier(sdk, this.bootstrapStackName);
+
+    const activeAssets = new ActiveAssetCache();
+
+    let refreshStacksRunning = false;
+    const refreshStacks = async (isInitial?: boolean) => {
+      if (refreshStacksRunning) {
+        return;
+      }
+
+      refreshStacksRunning = true;
+
+      try {
+        const stacks = await this.fetchAllStackTemplates(cfn, qualifier);
+        for (const stack of stacks) {
+          activeAssets.rememberStack(stack);
+        }
+      } catch (err) {
+        throw new Error(`Error refreshing stacks: ${err}`);
+      } finally {
+        refreshStacksRunning = false;
+
+        if (!isInitial) {
+          setTimeout(refreshStacks, 300_000);
+        }
+      }
+    };
+
+    // Grab stack templates first
+    await refreshStacks(true);
+    // Refresh stacks in the background
+    const timeout = setTimeout(refreshStacks, 300_000);
+
+    try {
+      const bucket = await this.bootstrapBucketName(sdk, this.bootstrapStackName);
+      const numObjects = await this.numObjectsInBucket(s3, bucket);
+      const batches = 1;
+      const batchSize = 1000;
+      const currentTime = Date.now();
+      const graceDays = this.props.rollbackBufferDays;
+
+      print(chalk.white(`Parsing through ${numObjects} objects in batches`));
+
+      // Process objects in batches of 1000
+      // This is the batch limit of s3.DeleteObject and we intend to optimize for the "worst case" scenario
+      // where gc is run for the first time on a long-standing bucket where ~100% of objects are isolated.
+      for await (const batch of this.readBucketInBatches(s3, bucket, batchSize)) {
```

Comment:
I just thought of something else: we should ignore objects that have been written since the last refresh of the stack templates.

That'll prevent the following thing from happening:

```
              <--- collect templates
upload assets
              <--- GC HERE
deploy stack  <--- oops assets gone
```

Reply:
was hoping that `listObjectsV2` would take a snapshot of the bucket or something, but that was wishful thinking: https://stackoverflow.com/questions/68532631/is-amazon-s3s-listobjectsv2-self-consistent-over-multiple-pages

Reply:
now ignoring assets that are deleted after gc starts